vllm/inputs/data.py

```diff
@@ -122,6 +122,12 @@ class LLMInputs(TypedDict):
     if the model supports it.
     """
 
+    multi_modal_placeholders: NotRequired[
+        Optional["MultiModalPlaceholderDict"]]
+    """
+    Placeholder ranges for the multi-modal data.
+    """
```

Comment:
Since this field always goes with `multi_modal_data`, I suggest adding a level of nesting to this data structure like so:

```py
class MultiModalInputs(TypedDict):
    data: "MultiModalDataDict"
    placeholders: "MultiModalPlaceholderDict"

class LLMInputs(TypedDict):
    prompt_token_ids: List[int]
    prompt: NotRequired[Optional[str]]
    multi_modal_inputs: NotRequired[Optional[MultiModalInputs]]
```

This should avoid the need to perform an extra `if` check inside each model's input processor.

Reply:
I'm happy to do that but I think the signature ends up being

```
class MultiModalInputs(TypedDict):
    data: "MultiModalDataDict"
    placeholders: Optional["MultiModalPlaceholderDict"]
```

so the check for placeholders still seems necessary.

That being said, I was previously under the mistaken impression that `LLMInputs` was part of the external API and so consumers could explicitly specify the placeholder locations to skip the processor but I see now that that's not really the case, at least not without changing `TextPrompt`/`TokensPrompt`. Which then begs a few questions:

- Should the placeholder dict be added to the external API as well?
- If so, is it worth making a breaking change to `TextPrompt`/`TokensPrompt`?
- If not, should `LLMInputs` be consistent with the external APIs?

(Additionally, if there's no practical way to explicitly specify the placeholders anyway I can simply remove the check and have the processors ignore that possibility.)

Reply:
> so consumers could explicitly specify the placeholder locations to skip the processor

Is there a particular use case for this? I think it'll just end up complicating the code.

Reply:
Not concretely, it was more just that because the processor is `LLMInputs -> LLMInputs` it seemed like "don't mess with placeholder annotations if they already exist" would be the most sensible semantics if they _were_ already provided.

But I'm happy to either assert that they don't already exist or blindly replace them instead -- let me know.

Reply:
I see. We can work on this in another PR then, since I also have another PR that refactors the input structure. (#8688)